feat: better in-app documentation (#417)
knocklabs/docs: content/in-app-ui/api-overview.mdx

---
title: In-app APIs and real-time service
description: Learn more about the capabilities of Knock's in-app APIs and real-time services, and how these can power robust in-app notification experiences with little effort.
section: Building in-app UI
---

> Knock provides a complete set of APIs to power your in-app notifications when creating notifications for the Knock-powered in-app channel. These APIs cover:

Reviewer: This is a bit redundant. Maybe something like:

```suggestion
Knock provides a complete set of APIs to render your in-app notifications with the Knock-powered in-app channel. These APIs cover:
```

- Fetching a reverse chronological list of in-app feed messages for a user. ([API reference](/reference#get-feed)).
- Retrieving badge count information on the number of seen and unread messages a user has.
- Handling message engagement to mark messages as seen, read, and archived.
- Real-time delivery of in-app messages from Knock to your user.

Our client-side SDKs wrap these APIs and provide a convenient way to interact with the Knock APIs.

## Security model

Knock's in-app APIs are accessible via our client SDKs, which use your public API key to authenticate. In addition, we support an enhanced security mode that signs a request against the current user and provides an additional authorization mechanism. You can read more in [our security documentation](/in-app-ui/security-and-authentication).

## Real-time delivery

> Knock provides a real-time, web socket connection for your users to subscribe to new messages being produced on an in-app channel. The events we currently send over the websocket are:

Reviewer: ```suggestion
Knock provides a real-time web socket connection for your users to subscribe to new messages being produced on an in-app channel. The events we currently send over the websocket are:
```

Reviewer: In this case I just changed the second instance to match the first, but maybe we should decide how we want the usage to be across our docs for consistency. It looks like the official protocol is "WebSocket," so maybe "websocket" (all one word) is the way to go 🤷

| Event         | Description                                                                                                                                                                              |
| ------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `new-message` | Emitted when a new message is produced over an in-app channel. Will not contain the full message contents, but will include metadata about the updated badge counts. Might be throttled. |

## Filtering in-app notifications

> It's possible to pass filters to the in-app messages endpoint. This makes it possible to:

Reviewer: Slightly redundant again: 
```suggestion
It's possible to pass filters to the in-app messages endpoint. This allows you to:
```

- Only return in-app messages for a particular tenant
- Only return archived or unarchived messages

> - Filter messages by the `data` associated

Reviewer 2: ```suggestion
- Filter messages by the `data` payload used to trigger the notification
```

> You can see the full set of available filters in the [endpoint documentation](/reference#get-feed). Please note, by default all `archived` messages are not displayed.

Reviewer: >Please note, by default all `archived` messages are not displayed.

I wonder if this warrants a callout block, or at least some bolded text to make it more eye-catching? 

>**Please note:** by default, `archived` messages are not displayed.

## In-app API response

### `FeedMetadata`

<Attributes>
  <Attribute
    name="total_count"
    type="number"
    description="The total number of messages the user has for this in-app channel."
  />
  <Attribute
    name="unread_count"
    type="number"
    description="The total number of unread items the user has on the in-app channel."
  />
  <Attribute
    name="unseen_count"
    type="number"
    description="The total number of unseen items the user has on the in-app channel."
  />
</Attributes>

### `FeedItem`

> Requests to the feed endpoint will return a `FeedItem`. Each feed item includes:

Reviewer 2: An array of FeedItems?

  />
  <Attribute
    name="source"
    type="WorkflowSource"
    description="The key, version_id, and categories of the workflow that generated this in-app message."
  />
  <Attribute
    name="tenant"
    type="string"
    description="An optional tenant identifier that was set in the workflow run scope."
  />
  <Attribute
    name="archived_at"
    type="utc_datetime"
    description="An optional tenant identifier that was set in the workflow run scope."
  />
  <Attribute
    name="read_at"
    type="utc_datetime"
    description="When set, indicates the last time the message was marked as read."
  />
  <Attribute
    name="seen_at"
    type="utc_datetime"
    description="When set, indicates the last time the message was marked as seen."
  />
  <Attribute
    name="link_clicked_at"
    type="utc_datetime"
    description="When set, indicates the last time the message was clicked."
  />
  <Attribute
    name="inserted_at"
    type="utc_datetime"
    description="The time the message was generated."
  />
  <Attribute
    name="updated_at"
    type="utc_datetime"
    description="The time the message was last updated."
  />
</Attributes>

### Message contents (`ContentBlock`)

> Each in-app message will contain the contents of the underlying in-app message template that was used to generate the message, which you can use to display your in-app notifications which you'll find under the `blocks` attribute on the `FeedItem` returned. Each block includes:

Reviewer 2: ```suggestion
Each in-app message will contain the contents of the message template that was used to generate the message, which you can use to display your in-app notifications. You'll find this content under the `blocks` attribute on each `FeedItem` returned. Each block includes:
```

<Attributes>
  <Attribute
    name="key"
    type="string"
    description="A unique key for the block."
  />
  <Attribute
    name="type"
    type="markdown | string"
    description="The type of block."
  />
  <Attribute
    name="rendered"
    type="string"
    description="The complete rendered content for the block."
  />
  <Attribute
    name="content"
    type="string"
    description="The liquid template associated with the block."
  />
</Attributes>

**Note**: when the block type is `markdown`, the `rendered` content will always be HTML that should be rendered for the in-app message.

### Passing through data

> When triggering a workflow that generates an in-app message, by default all of the `data` in the workflow run scope will be embedded into the generated in-app message. That makes it possible to pass through identifiers from your system to your in-app notifications that you can use. Here's an example: if I triggered a workflow with `{ "project_id": "proj_123" }` which generated an in-app message, then I can expect the `data` in the `FeedItem` to include this project_id.

Reviewer: Small additions to @JEverhart383's suggestion
```suggestion
When triggering a workflow that generates an in-app message, all of the `data` in the workflow run scope will be included in the generated in-app message payload. This makes it possible to pass through identifiers from your system to your in-app notifications that you can then use to render custom UI elements.
 
Here's an example: if you trigger a workflow with `{ "project_id": "proj_123" }` which generates an in-app message, you can expect the `data` in the `FeedItem` to include this `project_id`.
```

I'm curious if this is actually a "default" behavior (which implies it can be modified) or if it's just the way it works (which I think is the case). If the latter, I would remove "by default" altogether.
